src/jit/gentree.h

```diff
@@ -574,11 +574,21 @@ struct GenTree
         return isContained() && IsCnsIntOrI() && !isUsedFromSpillTemp();
     }
 
+#if FEATURE_X87_DOUBLES
```

Comment:
For the former, just the recommendations specified in Intel Architecture Optimization Manual: http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-optimization-manual.pdf and in the AMD Software Optimization Guide: https://support.amd.com/TechDocs/47414_15h_sw_opt_guide.pdf.

For the second, a very simple example is `1.70141173E+38f * 3 / 2`. When done using purely single-precision arithmetic, the result is `PositiveInfinity`, however when each constant is first cast to double and then downcast again `(float)((double)1.70141173E+38f * (double)3 / (double)2)` you get a different result: `2.55211755E+38`. This is also where order of operations matters, because `1.70141173E+38f * 1.5f` also returns `2.55211755E+38` rather than `PositiveInfinity`.

Most of the constant folding around float/double appears to be disable today due to the issues that can arise if things are done wrong (and because we don't allow carrying of single-precision constants).